numba/np/arraymath.py

```diff
@@ -864,6 +864,98 @@ def flat_all(a):
     return flat_all
 
 
+@register_jitable
+def _can_broadcast_returns_bool(array, dest_shape):
+    src_shape = array.shape
+    src_ndim = len(src_shape)
+    dest_ndim = len(dest_shape)
+    if src_ndim > dest_ndim:
+        return False
+    for size in dest_shape:
+        if size < 0:
+            return False
+
+    # based on _broadcast_onto function in numba/np/npyimpl.py
+    src_index = 0
+    dest_index = dest_ndim - src_ndim
+    while src_index < src_ndim:
+        src_dim = src_shape[src_index]
+        dest_dim = dest_shape[dest_index]
+        # possible cases for (src_dim, dest_dim):
+        #  * (1, 1)   -> Ok
+        #  * (>1, 1)  -> Error!
+        #  * (>1, >1) -> src_dim == dest_dim else error!
+        #  * (1, >1)  -> Ok
+        if src_dim == dest_dim or src_dim == 1:
+            src_index += 1
+            dest_index += 1
+        else:
+            return False
+
+    return True
+
+
+@overload(np.allclose)
+@overload_method(types.Array, "allclose")
+def np_allclose(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
+
+    if not type_can_asarray(a):
+        raise TypeError('The first argument "a" must be array-like')
+
+    if not type_can_asarray(b):
+        raise TypeError('The first argument "b" must be array-like')
+
+    if not isinstance(rtol, types.Float):
+        raise TypeError('The fourth argument "rtol" must be a '
+                        'floating point')
+
+    if not isinstance(atol, types.Float):
+        raise TypingError('The fifth argument "atol" must be a '
+                          'floating point')
+
+    if not isinstance(equal_nan, types.Boolean):
+        raise TypeError('The sixth argument "equal_nan" must be a '
+                        'boolean')
+
+    def np_allclose_impl(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
+        a = np.asarray(a)
+        b = np.asarray(b)
+        if a.shape != b.shape:
+            if _can_broadcast_returns_bool(a, b.shape):
+                a = np.broadcast_to(a, b.shape)
+            elif _can_broadcast_returns_bool(b, a.shape):
+                b = np.broadcast_to(b, a.shape)
+            else:
+                raise ValueError('operands could not be broadcast together '
+                                 'with remapped shapes')
```

Comment:
The broadcasting here fails if one of `a` or `b` is a scalar. The reason being, `.shape` will be `()`(an empty `UniTuple`) and it will fail while JITing the code.

Reply:
As you notice, one cannot call `np.broadcast_to` with an empty shape. In the case one of the arguments is an integer, you can iterate over the array and check if they're the same:

```python
if isinstance(b, types.Integer):
    def impl(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
        for av in np.nditer(a):
            if not allclose_inner(av, b, rtol, atol, equal_nan):
                return False
            return True
    return impl
else:
    def np_allclose_impl(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
        ...
    return np_allclose_impl
```

You can replicate the idea above for the cases where
* both `a` and `b` are scalars
* either `a` or `b` is scalar
* both `a` and `b` are arrays

Reply:
This sounds like a nice approach. Thanks for the idea.